docs/internals/sorting.md

```diff
@@ -0,0 +1,93 @@
+# Sorting
+
+Quickwit can sort results based on fastfield values or score. This document discuss where and ho
+ it happens.
+It also tries to describe optimizations that may be enabled (but are not necessarily implemente)
+by this behavior.
+
+# Current state
+(this section shall be removed once proposed changes are accepted)
+
+## Behavior
+
+Sorting is controlled by the `sort_by` query parameter. It accepts a comma separated list of fields
+to use for sorting. Sorting is Ascending by default. The sorting order can be reversed by prefixing
+a field name with a hyphen `-`.
+The special value `_score` means sorting by score, with a sort that is then Descending by default
+(documents matching better goes first).
+
+In case of equality between two documents, the GlobalDocId, composed of (SplitId, SegmentId, DocId)
+is used as a tie breaker. It is always used in Ascending order *inside a split*. It is used in the
+same order as the first field *when merging results*
+// TODO is this really the case? That's at least what my own tests seems to show.
+
+If a document doesn't have a value for a sorting field, that document is considered to go after any
+document which has a value, independantly of sort order. That is, when sorting the value 1,2 and
+None, ascending sort would give `[1, 2, None]`, and descending sort would give `[2, 1, None]`.
+
+If the client does not request sorting, documents are sorted using (DocId, SplitId, SegmentId), in
+Descending order.
+
+There are also plain bugs in sorting, in particular when values are absent (like 
+https://github.com/quickwit-oss/quickwit/blob/5d2c66e38e17ba5952196f01123d87f5a98b6a3c/quickwit/quickwit-search/src/collector.rs#L379-L380
+which prevent further admission in top-k once a single document with an absent value makes it into
+the top-k, such as if it's one of the first matching documents in the split).
+
+## Code
+Sorting happends in multiple places. First in `QuickwitSegmentCollector::collect_top_k`, for
+getting top documents from a single segment (and a single split). Then in 
+`top_k_partial_hits` when merging splits inside a single leaf node. The retry mechanism in
+cluster\_client can also end up merging results in `merge_leaf_search_response`, this time
+only concatenating partial hits. Finally results from splits on different leaf nodes are
+merged using the same logic as merging splits inside a leaf node, executed on the root node.
+
+# Proposed state
+
+## Behavior
+
+Sorting is controlled by the `sort_by` query parameter. It accepts a comma separated list of fields
+to use for sorting. Sorting is Ascending by default. The sorting order can be reversed by prefixing
+a field name with a hyphen `-`.
+The special value `_score` means sorting by score, with a sort that is then Descending by default
+(documents matching better goes first).
+
+In case of equality between two documents, the GlobalDocId, composed of (SplitId, SegmentId, DocId)
```

Comment:
Hey :wave:,
This might be based on wrong assumptions, disregard if so :D
With [cursor pagination](https://betterprogramming.pub/understanding-the-offset-and-cursor-pagination-8ddc54d10d98?gi=72175bea8ea5) in mind: shouldn't the `timestamp` be somewhere in the tie breaker, espacially if it's the posting's sort order? It probably is as it seems like an optim for top-k on your prefered "timestamp" sort?

Reply:
I think if you want `timestamp` to be used as a tie breaker of sorts, you should put it last in the `sort_by` query parameter. Right now it's limited to two fields, but that limitation will be lifted off at some point. I don't think it would help in optimization to have `timestamp` as a tie breaker, on the contrary it means you have to always download that fastfield. However a "more stable sort order" is something that would be nice to have